### PR TITLE
TEZ-2119: Counter for launched containers

### PR DESCRIPTION
**Pull request by abstractdog**
All the changes are according to the conversation on Jira + addressing comments on this PR so far.
Added comments myself to clarify what unrelated changes mean

### CODE REVIEW

File: tez-api/src/main/java/org/apache/tez/serviceplugins/api/TaskSchedulerContext.java [hunk 1]
```diff
@@ -84,6 +84,20 @@ void taskAllocated(Object task,
...
+   *
+   * @param the actual container
+   */
+  void containerAllocated(Container container);
+
+  /**
+   * Indicate to the framework that a container is being reused:
+   * there is a task assigned to an already used container.
+   *
+   * @param the actual container
```
Comment (ayushtkn) [inline tez-api/src/main/java/org/apache/tez/serviceplugins/api/TaskSchedulerContext.java:98]: should be 
```
   * @param container the actual container
```

File: tez-dag/src/main/java/org/apache/tez/dag/app/DAGAppMaster.java [hunk 2]
```diff
@@ -895,6 +898,12 @@ protected synchronized void handle(DAGAppMasterEvent event) {
     }
   }
 
+  private void handleUsedContainersOnDagFinish(DAG dag) {
+    dag.incrementDagCounter(DAGCounter.TOTAL_CONTAINERS_USED,
```
Comment (ayushtkn) [inline tez-dag/src/main/java/org/apache/tez/dag/app/DAGAppMaster.java:902]: is it actually an increment or a set? mens just from the name point of view + can this have some leftovers due to failures + retires and stuff like that?

File: tez-dag/src/main/java/org/apache/tez/dag/app/rm/TaskSchedulerManager.java [hunk 3]
```diff
@@ -951,6 +954,14 @@ public void dagCompleted() {
     }
   }
 
+  public int getAlreadyHeldContainersCount() {
+    int count = 0;
+    for (int i = 0 ; i < taskSchedulers.length ; i++) {
+      count += taskSchedulers[i].getTaskScheduler().getHeldContainersCount();
+    }
```
Comment (ayushtkn) [inline tez-dag/src/main/java/org/apache/tez/dag/app/rm/TaskSchedulerManager.java:961]: enhanced for loop looks more fancy
```
    for (TaskSchedulerWrapper taskScheduler : taskSchedulers) {
      count += taskScheduler.getTaskScheduler().getHeldContainersCount();
    }
```

File: tez-dag/src/main/java/org/apache/tez/dag/app/DAGAppMaster.java (hunk 2 above)
Comment (abstractdog) [reply tez-dag/src/main/java/org/apache/tez/dag/app/DAGAppMaster.java:902]: yeah, idempotency is always better, I'm simply creating a setCounter in DAGImpl

File: tez-api/src/main/java/org/apache/tez/serviceplugins/api/TaskSchedulerContext.java (hunk 1 above)
Comment (abstractdog) [reply tez-api/src/main/java/org/apache/tez/serviceplugins/api/TaskSchedulerContext.java:98]: ack

File: tez-api/src/main/java/org/apache/tez/serviceplugins/api/TaskSchedulerContext.java [hunk 4]
```diff
@@ -84,6 +84,20 @@ void taskAllocated(Object task,
                      Object appCookie,
                      Container container);
 
+  /**
+   * Indicate to the framework that a container is being allocated.
+   *
+   * @param the actual container
```
[the remark this answers is not in the document]
Comment (abstractdog) [reply tez-api/src/main/java/org/apache/tez/serviceplugins/api/TaskSchedulerContext.java:90]: ack

File: tez-api/src/main/java/org/apache/tez/serviceplugins/api/TaskScheduler.java [hunk 5]
```diff
@@ -263,4 +265,19 @@ public abstract boolean deallocateTask(Object task, boolean taskSucceeded,
    */
   public abstract void dagComplete() throws ServicePluginException;
 
+  /**
+   * Get the number of held containers
+   * @throws ServicePluginException when the service runs into a fatal error which it cannot handle.
+   *                               This will cause the app to shutdown.
+   */
+  public abstract int getHeldContainersCount();
```
[the remark this answers is not in the document]
Comment (abstractdog) [reply tez-api/src/main/java/org/apache/tez/serviceplugins/api/TaskScheduler.java:273]: eh, yeah, a leftover, cleaning up

**Comment by abstractdog** [comment]
> Thanx @abstractdog for the PR. Sounds cool, minor stuff. Do you plan to extend a test to cover this?

yes, absolutely, I'm trying to add a UT

**Comment by tez-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |  26m 12s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 4 new or modified test files.  |
||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   5m 50s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   9m 55s |  master passed  |
| +1 :green_heart: |  compile  |   2m  5s |  master passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu122.04.1  |
| +1 :green_heart: |  compile  |   1m 57s |  master passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~22.04-b09  |
| +1 :green_heart: |  checkstyle  |   2m  8s |  master passed  |
| +1 :green_heart: |  javadoc  |   1m 50s |  master passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu122.04.1  |
| +1 :green_heart: |  javadoc  |   1m 38s |  master passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~22.04-b09  |
| +0 :ok: |  spotbugs  |   0m 49s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   3m 43s |  master passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 10s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   1m 11s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 18s |  the patch passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu122.04.1  |
| +1 :green_heart: |  javac  |   1m 18s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m  4s |  the patch passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~22.04-b09  |
| +1 :green_heart: |  javac  |   1m  4s |  the patch passed  |
| -0 :warning: |  checkstyle  |   0m 13s |  tez-api: The patch generated 1 new + 16 unchanged - 0 fixed = 17 total (was 16)  |
| -0 :warning: |  checkstyle  |   0m 32s |  tez-dag: The patch generated 5 new + 413 unchanged - 0 fixed = 418 total (was 413)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| -1 :x: |  javadoc  |   0m 25s |  tez-api-jdkUbuntu-11.0.19+7-post-Ubuntu-0ubuntu122.04.1 with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu122.04.1 generated 2 new + 13 unchanged - 0 fixed = 15 total (was 13)  |
| -1 :x: |  javadoc  |   0m 26s |  tez-api-jdkPrivateBuild-1.8.0_362-8u372-ga~us1-0ubuntu1~22.04-b09 with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~22.04-b09 generated 2 new + 0 unchanged - 0 fixed = 2 total (was 0)  |
| +1 :green_heart: |  findbugs  |   2m 49s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   2m 18s |  tez-api in the patch passed.  |
| -1 :x: |  unit  |   5m 15s |  tez-dag in the patch failed.  |
| +1 :green_heart: |  unit  |   4m 11s |  tez-ext-service-tests in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 34s |  The patch does not generate ASF License warnings.  |
|  |   |  78m 31s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | tez.dag.app.rm.TestContainerReuse |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-301/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/tez/pull/301 |
| JIRA Issue | TEZ-2119 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs checkstyle compile |
| uname | Linux 85e8f1818277 4.15.0-212-generic #223-Ubuntu SMP Tue May 23 13:09:22 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/tez.sh |
| git revision | master / 50380751b |
| Default Java | Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~22.04-b09 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu122.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~22.04-b09 |
| checkstyle | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-301/1/artifact/out/diff-checkstyle-tez-api.txt |
| checkstyle | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-301/1/artifact/out/diff-checkstyle-tez-dag.txt |
| javadoc | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-301/1/artifact/out/diff-javadoc-javadoc-tez-api-jdkUbuntu-11.0.19+7-post-Ubuntu-0ubuntu122.04.1.txt |
| javadoc | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-301/1/artifact/out/diff-javadoc-javadoc-tez-api-jdkPrivateBuild-1.8.0_362-8u372-ga~us1-0ubuntu1~22.04-b09.txt |
| unit | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-301/1/artifact/out/patch-unit-tez-dag.txt |
|  Test Results | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-301/1/testReport/ |
| Max. process+thread count | 1014 (vs. ulimit of 5500) |
| modules | C: tez-api tez-dag tez-ext-service-tests U: . |
| Console output | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-301/1/console |
| versions | git=2.34.1 maven=3.6.3 findbugs=3.0.1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by tez-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |  31m  4s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 7 new or modified test files.  |
||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   6m  4s |  Maven dependency ordering for branch  |
| -1 :x: |  mvninstall  |   9m 57s |  root in master failed.  |
| +1 :green_heart: |  compile  |   1m 37s |  master passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu122.04.1  |
| +1 :green_heart: |  compile  |   1m 16s |  master passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~22.04-b09  |
| +1 :green_heart: |  checkstyle  |   1m 31s |  master passed  |
| +1 :green_heart: |  javadoc  |   1m 19s |  master passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu122.04.1  |
| +1 :green_heart: |  javadoc  |   0m 53s |  master passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~22.04-b09  |
| +0 :ok: |  spotbugs  |   0m 38s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   3m 31s |  master passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 10s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   1m 26s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 36s |  the patch passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu122.04.1  |
| +1 :green_heart: |  javac  |   1m 36s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 22s |  the patch passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~22.04-b09  |
| +1 :green_heart: |  javac  |   1m 22s |  the patch passed  |
| -0 :warning: |  checkstyle  |   0m 35s |  tez-dag: The patch generated 11 new + 555 unchanged - 6 fixed = 566 total (was 561)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  javadoc  |   0m 52s |  the patch passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu122.04.1  |
| +1 :green_heart: |  javadoc  |   0m 50s |  the patch passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~22.04-b09  |
| +1 :green_heart: |  findbugs  |   3m 25s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   2m 21s |  tez-api in the patch passed.  |
| -1 :x: |  unit  |   5m 14s |  tez-dag in the patch failed.  |
| +1 :green_heart: |  unit  |   4m  5s |  tez-ext-service-tests in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 31s |  The patch does not generate ASF License warnings.  |
|  |   |  81m 38s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | tez.dag.app.rm.TestContainerReuse |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-301/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/tez/pull/301 |
| JIRA Issue | TEZ-2119 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs checkstyle compile |
| uname | Linux 15e95823fd8c 4.15.0-212-generic #223-Ubuntu SMP Tue May 23 13:09:22 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/tez.sh |
| git revision | master / 50380751b |
| Default Java | Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~22.04-b09 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu122.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~22.04-b09 |
| mvninstall | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-301/4/artifact/out/branch-mvninstall-root.txt |
| checkstyle | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-301/4/artifact/out/diff-checkstyle-tez-dag.txt |
| unit | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-301/4/artifact/out/patch-unit-tez-dag.txt |
|  Test Results | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-301/4/testReport/ |
| Max. process+thread count | 959 (vs. ulimit of 5500) |
| modules | C: tez-api tez-dag tez-ext-service-tests U: . |
| Console output | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-301/4/console |
| versions | git=2.34.1 maven=3.6.3 findbugs=3.0.1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by tez-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 35s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  1s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 7 new or modified test files.  |
||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   5m 57s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  11m 49s |  master passed  |
| +1 :green_heart: |  compile  |   2m  4s |  master passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu122.04.1  |
| +1 :green_heart: |  compile  |   1m 58s |  master passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~22.04-b09  |
| +1 :green_heart: |  checkstyle  |   2m 14s |  master passed  |
| +1 :green_heart: |  javadoc  |   1m 49s |  master passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu122.04.1  |
| +1 :green_heart: |  javadoc  |   1m 37s |  master passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~22.04-b09  |
| +0 :ok: |  spotbugs  |   0m 50s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   3m 44s |  master passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 10s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   1m  8s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 13s |  the patch passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu122.04.1  |
| +1 :green_heart: |  javac  |   1m 13s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m  7s |  the patch passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~22.04-b09  |
| +1 :green_heart: |  javac  |   1m  7s |  the patch passed  |
| -0 :warning: |  checkstyle  |   0m 36s |  tez-dag: The patch generated 5 new + 554 unchanged - 7 fixed = 559 total (was 561)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  javadoc  |   0m 47s |  the patch passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu122.04.1  |
| +1 :green_heart: |  javadoc  |   0m 46s |  the patch passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~22.04-b09  |
| +1 :green_heart: |  findbugs  |   2m 48s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   2m 15s |  tez-api in the patch passed.  |
| +1 :green_heart: |  unit  |   5m 12s |  tez-dag in the patch passed.  |
| +1 :green_heart: |  unit  |   4m 10s |  tez-ext-service-tests in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 34s |  The patch does not generate ASF License warnings.  |
|  |   |  54m 57s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-301/5/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/tez/pull/301 |
| JIRA Issue | TEZ-2119 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs checkstyle compile |
| uname | Linux f51053af8b3f 4.15.0-212-generic #223-Ubuntu SMP Tue May 23 13:09:22 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/tez.sh |
| git revision | master / 50380751b |
| Default Java | Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~22.04-b09 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu122.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~22.04-b09 |
| checkstyle | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-301/5/artifact/out/diff-checkstyle-tez-dag.txt |
|  Test Results | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-301/5/testReport/ |
| Max. process+thread count | 1012 (vs. ulimit of 5500) |
| modules | C: tez-api tez-dag tez-ext-service-tests U: . |
| Console output | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-301/5/console |
| versions | git=2.34.1 maven=3.6.3 findbugs=3.0.1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by abstractdog** [comment]
finally a working version for this patch, unit tests are included, tested on cluster as show below:

1. ran a dag with resources available for 2 tasks containers (after the first 2 allocation/launch, the rest 6 attempts were all reuse as expected)
```
INFO  : org.apache.tez.common.counters.DAGCounter:
INFO  :    NUM_SUCCEEDED_TASKS: 8
INFO  :    TOTAL_LAUNCHED_TASKS: 8
INFO  :    OTHER_LOCAL_TASKS: 5
INFO  :    AM_CPU_MILLISECONDS: 5130
INFO  :    AM_GC_TIME_MILLIS: 141
INFO  :    INITIAL_HELD_CONTAINERS: 0
INFO  :    TOTAL_CONTAINERS_USED: 2
INFO  :    TOTAL_CONTAINER_ALLOCATION_COUNT: 2
INFO  :    TOTAL_CONTAINER_LAUNCH_COUNT: 2
INFO  :    TOTAL_CONTAINER_REUSE_COUNT: 6
```

2. ran the dag quickly again: the same amount of tasks, but already had 2 containers, so all 8 attempts were running on reused containers (on the same 2 containers as shown in TOTAL_CONTAINERS_USED)
```
INFO  : org.apache.tez.common.counters.DAGCounter:
INFO  :    NUM_SUCCEEDED_TASKS: 8
INFO  :    TOTAL_LAUNCHED_TASKS: 8
INFO  :    OTHER_LOCAL_TASKS: 5
INFO  :    AM_CPU_MILLISECONDS: 1130
INFO  :    AM_GC_TIME_MILLIS: 83
INFO  :    INITIAL_HELD_CONTAINERS: 2
INFO  :    TOTAL_CONTAINERS_USED: 2
INFO  :    TOTAL_CONTAINER_REUSE_COUNT: 8
```

3. killed another yarn application to make resources, this is the same as 1) just with 3 containers:
```
INFO  : org.apache.tez.common.counters.DAGCounter: 
INFO  :    NUM_SUCCEEDED_TASKS: 8
INFO  :    TOTAL_LAUNCHED_TASKS: 8
INFO  :    OTHER_LOCAL_TASKS: 5
INFO  :    AM_CPU_MILLISECONDS: 5790
INFO  :    AM_GC_TIME_MILLIS: 184
INFO  :    INITIAL_HELD_CONTAINERS: 0
INFO  :    TOTAL_CONTAINERS_USED: 3
INFO  :    TOTAL_CONTAINER_ALLOCATION_COUNT: 3
INFO  :    TOTAL_CONTAINER_LAUNCH_COUNT: 3
INFO  :    TOTAL_CONTAINER_REUSE_COUNT: 5
```

4. ran dag again, the same as 2) but with 3 containers:
```
INFO  : org.apache.tez.common.counters.DAGCounter:
INFO  :    NUM_SUCCEEDED_TASKS: 8
INFO  :    TOTAL_LAUNCHED_TASKS: 8
INFO  :    OTHER_LOCAL_TASKS: 5
INFO  :    AM_CPU_MILLISECONDS: 1360
INFO  :    AM_GC_TIME_MILLIS: 77
INFO  :    INITIAL_HELD_CONTAINERS: 3
INFO  :    TOTAL_CONTAINERS_USED: 3
INFO  :    TOTAL_CONTAINER_REUSE_COUNT: 8
```

File: tez-dag/src/test/java/org/apache/tez/dag/app/rm/TestContainerReuse.java [hunk 6]
```diff
@@ -1112,8 +1137,8 @@ public void testReuseConflictLocalResources() throws IOException, InterruptedExc
     taskSchedulerManager.init(tezConf);
     taskSchedulerManager.start();
 
-    TaskSchedulerWithDrainableContext taskScheduler = (TaskSchedulerWithDrainableContext) ((TaskSchedulerManagerForTest) taskSchedulerManager)
-        .getSpyTaskScheduler();
+    TaskSchedulerWithDrainableContext taskScheduler =
```
Comment (abstractdog) [inline tez-dag/src/test/java/org/apache/tez/dag/app/rm/TestContainerReuse.java:1140]: only a reformat

File: tez-dag/src/test/java/org/apache/tez/dag/app/rm/TestTaskSchedulerManager.java [hunk 7]
```diff
@@ -175,7 +175,7 @@ protected void notifyForTest() {
   DAGClientServer mockClientService;
   TestEventHandler mockEventHandler;
   ContainerSignatureMatcher mockSigMatcher;
-  MockTaskSchedulerManager schedulerHandler;
```
Comment (abstractdog) [inline tez-dag/src/test/java/org/apache/tez/dag/app/rm/TestTaskSchedulerManager.java:178]: schedulerHandler was incorrect, simply renamed

File: tez-dag/src/test/java/org/apache/tez/dag/app/rm/TestTaskSchedulerManager.java [hunk 8]
```diff
@@ -975,14 +1003,15 @@ TaskScheduler createYarnTaskScheduler(TaskSchedulerContext taskSchedulerContext,
...
       testTaskSchedulers.add(yarnTaskScheduler);
       yarnSchedulerCreated.set(true);
+      when(yarnTaskScheduler.getHeldContainersCount()).thenReturn(YARN_TASK_SCHEDULER_HELD_CONTAINERS);
       return yarnTaskScheduler;
     }
 
     @Override
     TaskScheduler createUberTaskScheduler(TaskSchedulerContext taskSchedulerContext, int schedulerId) {
       taskSchedulerContexts.add(taskSchedulerContext);
+      testTaskSchedulers.add(uberTaskScheduler);
```
Comment (abstractdog) [inline tez-dag/src/test/java/org/apache/tez/dag/app/rm/TestTaskSchedulerManager.java:1013]: this doesn't have to do anything with this patch, it's just fixed for clarity's sake

File: tez-dag/src/test/java/org/apache/tez/dag/app/launcher/TestContainerLauncherManager.java [hunk 9]
```diff
@@ -514,26 +549,41 @@ public void stopContainer(ContainerStopRequest stopRequest) {
...
-          .reportError(ServicePluginErrorDefaults.SERVICE_UNAVAILABLE, "ReportError", new DagInfoImplForTest(DAG_INDEX, DAG_NAME));
+      getContext().reportError(ServicePluginErrorDefaults.SERVICE_UNAVAILABLE, "ReportError",
+          new DagInfoImplForTest(DAG_INDEX, DAG_NAME));
     }
   }
 
+  /**
+   * This container launcher simply implements ContainerLauncher methods with the proper context callbacks.
+   */
+  public static class ContainerLauncherForTest extends ContainerLauncher {
```
Comment (abstractdog) [inline tez-dag/src/test/java/org/apache/tez/dag/app/launcher/TestContainerLauncherManager.java:573]: originally: ContainerLauncherForTest was a mock launcher for reporting errors, I simply renamed it to FailureReporterContainerLauncher, and turned ContainerLauncherForTest to a NO-OP0-ish container launcher with simply callbacks

**Comment by tez-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 36s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  1s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 7 new or modified test files.  |
||| _ master Compile Tests _ |
| +0 :ok: |  mvndep  |   6m 10s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  12m  3s |  master passed  |
| +1 :green_heart: |  compile  |   2m  2s |  master passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu122.04.1  |
| +1 :green_heart: |  compile  |   1m 58s |  master passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~22.04-b09  |
| +1 :green_heart: |  checkstyle  |   2m 12s |  master passed  |
| +1 :green_heart: |  javadoc  |   1m 51s |  master passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu122.04.1  |
| +1 :green_heart: |  javadoc  |   1m 34s |  master passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~22.04-b09  |
| +0 :ok: |  spotbugs  |   0m 49s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   3m 46s |  master passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 10s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   1m  8s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 14s |  the patch passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu122.04.1  |
| +1 :green_heart: |  javac  |   1m 14s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m  7s |  the patch passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~22.04-b09  |
| +1 :green_heart: |  javac  |   1m  7s |  the patch passed  |
| -0 :warning: |  checkstyle  |   0m 36s |  tez-dag: The patch generated 5 new + 554 unchanged - 7 fixed = 559 total (was 561)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  javadoc  |   0m 46s |  the patch passed with JDK Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu122.04.1  |
| +1 :green_heart: |  javadoc  |   0m 47s |  the patch passed with JDK Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~22.04-b09  |
| +1 :green_heart: |  findbugs  |   2m 50s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   2m 18s |  tez-api in the patch passed.  |
| +1 :green_heart: |  unit  |   5m 12s |  tez-dag in the patch passed.  |
| +1 :green_heart: |  unit  |   4m 11s |  tez-ext-service-tests in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 34s |  The patch does not generate ASF License warnings.  |
|  |   |  55m 27s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-301/6/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/tez/pull/301 |
| JIRA Issue | TEZ-2119 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs checkstyle compile |
| uname | Linux a5ab820d4d95 4.15.0-212-generic #223-Ubuntu SMP Tue May 23 13:09:22 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/tez.sh |
| git revision | master / 50380751b |
| Default Java | Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~22.04-b09 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.19+7-post-Ubuntu-0ubuntu122.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~22.04-b09 |
| checkstyle | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-301/6/artifact/out/diff-checkstyle-tez-dag.txt |
|  Test Results | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-301/6/testReport/ |
| Max. process+thread count | 1015 (vs. ulimit of 5500) |
| modules | C: tez-api tez-dag tez-ext-service-tests U: . |
| Console output | https://ci-hadoop.apache.org/job/tez-multibranch/job/PR-301/6/console |
| versions | git=2.34.1 maven=3.6.3 findbugs=3.0.1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Review by jteagles — APPROVED** [review]
+1. Looks good and will be a great feature. Can you take care of committing this?

**Pull request merged**